test/e2e/storage/testsuites/testdriver.go

```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
 	// GetDriverInfo returns DriverInfo for the TestDriver
 	GetDriverInfo() *DriverInfo
 	// CreateDriver creates all driver resources that is required for TestDriver method
-	// except CreateVolume
+	// except CreateVolume. May be called more than once and should only do something on
+	// the first call.
 	CreateDriver()
-	// CreateDriver cleanup all the resources that is created in CreateDriver
+	// CreateDriver cleanup all the resources that is created in CreateDriver. There is
+	// no guarantee that CreateDriver succeeded or even was called at all, so the test driver
+	// has to track resources.
 	CleanupDriver()
-	// SkipUnsupportedTest skips test in Testpattern is not suitable to test with the TestDriver
-	SkipUnsupportedTest(testpatterns.TestPattern)
+}
+
+// FilterTestDriver is an optional interface that drivers can
+// implement to filter out unsuitable tests while tests get defined.
+type FilterTestDriver interface {
+	// IsTestSupported returns true if the Testpattern is
+	// suitable to test with the TestDriver. This will be called
+	// already while defining tests and unsupported tests will not even
+	// be added to the test suite.
+	IsTestSupported(testpatterns.TestPattern) bool
+}
+
+// BeforeEachTestDriver is an optional interface that drivers can
+// implement to hook into test execution. It can be used to initialize
```

Comment:
Do we need to allow this method to initialize additional resources?
Are there any existing use cases for it?

(I'm worrying about that it was due to my previous mistake in putting resource initialization code for GCE driver here, which has already been moved to right place in another commit.)

Reply:
It's not really needed and in fact later (https://github.com/kubernetes/kubernetes/pull/72002) I ended up removing it again because it didn't turn out to be that useful. CreateDriver gets called at the same time and can replace the extra `BeforeEachTestDriver`.

I'll remove it directly in this PR instead.

Reply:
Done.